src/sentry/snuba/query_subscription_consumer.py

```diff
@@ -58,42 +72,59 @@ def inner(func: TQuerySubscriptionCallable) -> TQuerySubscriptionCallable:
     return inner
 
 
-def parse_message_value(value: str) -> Dict[str, Any]:
+def parse_message_value(value: str, topic: str) -> PayloadV3:
     """
     Parses the value received via the Kafka consumer and verifies that it
     matches the expected schema.
     :param value: A json formatted string
     :return: A dict with the parsed message
     """
+    SUPPORTED_SCHEMA_VERSION = 3
+    old_version = False
+
     with metrics.timer("snuba_query_subscriber.parse_message_value.json_parse"):
-        wrapper: Dict[str, Any] = json.loads(value)
+        wrapper: SubscriptionResults = json.loads(value)
 
     with metrics.timer("snuba_query_subscriber.parse_message_value.json_validate_wrapper"):
         try:
-            jsonschema.validate(wrapper, SUBSCRIPTION_WRAPPER_SCHEMA)
-        except jsonschema.ValidationError:
-            metrics.incr("snuba_query_subscriber.message_wrapper_invalid")
-            raise InvalidSchemaError("Message wrapper does not match schema")
+            jsoncodec = JsonCodec(get_schema(topic)["schema"])
+            jsoncodec.validate(wrapper)
+        except ValidationError:
+            old_version = True
+            metrics.incr("snuba_query_subscriber.message_wrapper.old_validation")
+            try:
+                jsonschema.validate(wrapper, SUBSCRIPTION_WRAPPER_SCHEMA)
+            except jsonschema.ValidationError:
+                metrics.incr("snuba_query_subscriber.message_wrapper_invalid")
+                raise InvalidSchemaError("Message wrapper does not match schema")
 
     schema_version: int = wrapper["version"]
-    if schema_version not in SUBSCRIPTION_PAYLOAD_VERSIONS:
-        metrics.incr("snuba_query_subscriber.message_wrapper_invalid_version")
-        raise InvalidMessageError("Version specified in wrapper has no schema")
+    payload: PayloadV3 = wrapper["payload"]
+
+    if old_version:
+        if schema_version not in SUBSCRIPTION_PAYLOAD_VERSIONS:
+            metrics.incr("snuba_query_subscriber.message_wrapper_invalid_version")
+            raise InvalidMessageError("Version specified in wrapper has no schema")
+
+        with metrics.timer("snuba_query_subscriber.parse_message_value.json_validate_payload"):
+            try:
+                jsonschema.validate(payload, SUBSCRIPTION_PAYLOAD_VERSIONS[schema_version])
+            except jsonschema.ValidationError:
+                metrics.incr("snuba_query_subscriber.message_payload_invalid")
+                raise InvalidSchemaError("Message payload does not match schema")
+
+    else:
+        if schema_version != SUPPORTED_SCHEMA_VERSION:
+            metrics.incr("snuba_query_subscriber.message_wrapper_invalid_version")
+            raise InvalidMessageError("Version specified in wrapper has no schema")
```

Comment:
Isn't this already checked by the schema which defines this as a constant? Why do we need to check it again?

Reply:
The schema just defines `version` as an integer. If it's possible to set a required value in the schema I can change that, I'm only aware of required types/properties.

Reply:
Yeah, let's change it to a "const" there, and save ourselves some manual validation here?

Reply:
https://github.com/getsentry/sentry-kafka-schemas/pull/63 👍